File: python/secret_recipe_decoder.py
#!/usr/bin/python3
import pydoc
import sys
import os

# Caesar encoding, for use with decoding below
ENCODING = {
    'y': 'a',
    'h': 'b',
    'v': 'c',
    'x': 'd',
    'k': 'e',
    'p': 'f',
    'z': 'g',
    's': 'h',
    'a': 'i',
    'b': 'j',
    'e': 'k',
    'w': 'l',
    'u': 'm',
    'q': 'n',
    'n': 'o',
    'l': 'p',
    'm': 'q',
    'f': 'r',
    'o': 's',
    'i': 't',
    'g': 'u',
    'j': 'v',
    't': 'w',
    'd': 'x',
    'r': 'y',
    'c': 'z',
    '3': '0',
    '8': '1',
    '4': '2',
    '0': '3',
    '2': '4',
    '7': '5',
    '5': '6',
    '9': '7',
    '1': '8',
    '6': '9'
 }

"""An ingredient has an amount and a description.
For example: an Ingredient could have "1 cup" as the amount and "butter" as the description."""
class Ingredient():
    def __init__(self, amount, description) -> None:
        self.amount = amount
        self.description = description


def decode_string(str):
    """Given a string named str, use the Caesar encoding above to return the decoded string."""
    # TODO: implement me
    decoded = ""
    for char in str:
        if char in ENCODING:
            decoded += ENCODING[char]
        else:
            decoded+= char
    return decoded




def decode_ingredient(line):
    """Given an ingredient, decode the amount and description, and return a new Ingredient"""
    line = decode_string(line)
    ingredients = line.split("#")
    return Ingredient(ingredients[0], ingredients[1])



    



def main():
    """A program that decodes a secret recipe"""
    # TODO: implement me
    with open('decoded_recipe.txt', 'w') as output:
        with open("secret_recipe.txt", "r") as file:
            for line in file:
                ingredient = decode_ingredient(line)
                print(ingredient.amount + " " + ingredient.description, file=output, end="")

if __name__ == "__main__":
    main()

File: java/src/main/java/com/labs1904/SecretRecipeDecoder.java
package com.labs1904;


import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class SecretRecipeDecoder {
    private static Map<String, String> ENCODING = new HashMap<String, String>() {
        {
            put("y", "a");
            put("h", "b");
            put("v", "c");
            put("x", "d");
            put("k", "e");
            put("p", "f");
            put("z", "g");
            put("s", "h");
            put("a", "i");
            put("b", "j");
            put("e", "k");
            put("w", "l");
            put("u", "m");
            put("q", "n");
            put("n", "o");
            put("l", "p");
            put("m", "q");
            put("f", "r");
            put("o", "s");
            put("i", "t");
            put("g", "u");
            put("j", "v");
            put("t", "w");
            put("d", "x");
            put("r", "y");
            put("c", "z");
            put("3", "0");
            put("8", "1");
            put("4", "2");
            put("0", "3");
            put("2", "4");
            put("7", "5");
            put("5", "6");
            put("9", "7");
            put("1", "8");
            put("6", "9");
        }
    };

    /**
     * Given a string named str, use the Caesar encoding above to return the decoded string.
     * @param str
     * @return
     */
    public static String decodeString(String str) {

        StringBuilder emptyString = new StringBuilder();

        for (char s: str.toCharArray()) {
            String toChars = ENCODING.getOrDefault(String.valueOf(s),String.valueOf(s));
            emptyString.append(toChars);
        }
        return emptyString.toString();


    }

    /**
     * Given an ingredient, decode the amount and description, and return a new Ingredient
     * @param line
     * @return
     */
    public static Ingredient decodeIngredient(String line) {
       // return new Ingredient("1 cup", "butter");
        String[] splitter = line.split("#");
        String parts = decodeString(splitter[0]);
        String ingredient = decodeString(splitter[1]);

        return new Ingredient(parts, ingredient);
    }

    public static void main(String[] args) {
        try {
            List<String> decodedLines = Files.readAllLines(Paths.get("src/main/resources/secret_recipe.txt"))
                    .stream()
                    .filter(line -> !line.trim().isEmpty())
                    .map(line -> {
                        Ingredient decoded = decodeIngredient(line);
                        return decoded.getAmount() + " " + decoded.getDescription();
                    })
                    .collect(Collectors.toList());

            Files.write(Paths.get("decoded_recipe.txt"), decodedLines);
            System.out.println("Recipe decoded successfully!");

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
        }
    }
}
